cmd/s3-zip-handlers.go

```diff
@@ -0,0 +1,496 @@
+		return
+	}
+
+	if s3Error := checkRequestAuthType(ctx, r, policy.GetObjectAction, bucket, zipPath); s3Error != ErrNone {
+		if getRequestAuthType(r) == authTypeAnonymous {
+			// As per "Permission" section in
+			// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html
+			// If the object you request does not exist,
+			// the error Amazon S3 returns depends on
+			// whether you also have the s3:ListBucket
+			// permission.
+			// * If you have the s3:ListBucket permission
+			//   on the bucket, Amazon S3 will return an
+			//   HTTP status code 404 ("no such key")
+			//   error.
+			// * if you don’t have the s3:ListBucket
+			//   permission, Amazon S3 will return an HTTP
+			//   status code 403 ("access denied") error.`
+			if globalPolicySys.IsAllowed(policy.Args{
+				Action:          policy.ListBucketAction,
+				BucketName:      bucket,
+				ConditionValues: getConditionValues(r, "", "", nil),
+				IsOwner:         false,
+			}) {
+				_, err = getObjectInfo(ctx, bucket, zipPath, opts)
+				if toAPIError(ctx, err).Code == "NoSuchKey" {
+					s3Error = ErrNoSuchKey
+				}
+			}
+		}
+		writeErrorResponseHeadersOnly(w, errorCodes.ToAPIErr(s3Error))
+		return
+	}
+
+	var rs *HTTPRangeSpec
+
+	// Validate pre-conditions if any.
+	opts.CheckPrecondFn = func(oi ObjectInfo) bool {
+		return checkPreconditions(ctx, w, r, oi, opts)
+	}
+
+	zipObjInfo, err := getObjectInfo(ctx, bucket, zipPath, opts)
+	if err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
+
+	var zipInfo []byte
+
+	if z, ok := zipObjInfo.UserDefined[archiveInfoMetadataKey]; ok {
+		zipInfo = []byte(z)
+	} else {
+		zipInfo, err = updateObjectMetadataWithZipInfo(ctx, objectAPI, bucket, object)
+	}
+	if err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
+
+	file, err := zipindex.FindSerialized(zipInfo, object)
+	if err != nil {
+		if err == io.EOF {
+			writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrNoSuchKey), r.URL, guessIsBrowserReq(r))
+		} else {
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		}
+		return
+	}
+
+	objInfo := ObjectInfo{
+		Bucket:  bucket,
+		Name:    file.Name,
+		Size:    int64(file.UncompressedSize64),
+		ModTime: zipObjInfo.ModTime,
+	}
+
+	// Set standard object headers.
+	if err = setObjectHeaders(w, objInfo, nil, opts); err != nil {
+		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))
+		return
+	}
+
+	// Set any additional requested response headers.
+	setHeadGetRespHeaders(w, r.URL.Query())
+
+	// Successful response.
+	if rs != nil {
+		w.WriteHeader(http.StatusPartialContent)
+	} else {
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// Update the passed zip object metadata with the zip contents info, file name, modtime, size, etc..
+func updateObjectMetadataWithZipInfo(ctx context.Context, objectAPI ObjectLayer, bucket, object string) ([]byte, error) {
```

Comment:
You need to pass ObjectOptions such that if some one asks versionId we should return appropriate error such as version not found and also support versionId == null to return the `null` default version.

Reply:
> You need to pass ObjectOptions such that if some one asks versionId we should return appropriate error such as version not found and also support versionId == null to return the `null` default version.

The thing about updating metadata in versioning is that it creates a new version which is probably intrusive for the customer. Maybe we don't have to update in that case and we can check for the object last 1mb (in most cases) or less to fetch the list of files in the zip.

Reply:
Actually it looks like CopyObject in object layer already supports updating a specific version with metadata (not sure why in the first place, not sure if replication is related to this)

So I can use it. I will try it and update.

Reply:
It's related to replication